### PR TITLE
metrics: Add dockerfile for iperf3 tests

### LINKED ISSUE #553: metrics: Add dockerfile for iperf3 tests
Instead of using the image gabyct/network, have a dockerfile to have iperf3

### PR DESCRIPTION
**Pull request by GabyCT**
This will build an image with iperf3.

Fixes #553

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

File: metrics/network/iperf3_dockerfile/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+FROM fedora
+
+# Version of the Dockerfile
+LABEL DOCKERFILE_VERSION="1.0"
+
+# Install iperf3
+RUN dnf -y update && \
+    dnf -y install iperf3
+
+CMD ["/bin/bash"]
```
Comment (devimc) [inline metrics/network/iperf3_dockerfile/Dockerfile]: why bash and not iperf3 ?

Comment (GabyCT) [reply metrics/network/iperf3_dockerfile/Dockerfile]: @devimc , I replaced the `CMD` for `ENTRYPOINT` with `iperf3`, however, it have issues to run the client for example
```
$  docker run --runtime kata-runtime local-iperf3 'export TMPDIR=/dev/shm && iperf3 -c 172.17.0.2 -u -t 30'
iperf3: parameter error - must either be a client (-c) or server (-s)
```
The main issue seems that it does not accept the `export TMPDIR...` that we need to run the container

Comment (devimc) [reply metrics/network/iperf3_dockerfile/Dockerfile]: make sense, thanks @GabyCT

Comment (grahamwhaley) [reply metrics/network/iperf3_dockerfile/Dockerfile]: Also, sometimes we might want the container to be the server, and sometimes the client, so hard coding the `iperf3` commandline as the `CMD` might not scale out so well.

**Comment by grahamwhaley** [comment]
Hi @GabyCT The changes look fine, but when I tried to test it (I actually wanted to test the Dockerfile build as I have  small pending change locally to how we do that....), then I get a failure. Not every time, but I think every time it has to build the dockerfile at least.
If I run:
```sh
bash -x ./network-metrics-iperf3.sh -p
```

then I get
```sh
...
waiting for server...
++ count=0
++ sleep 0.5
++ '[' 0 -le 6 ']'
++ die 'iperf3 server init fails'
++ msg='iperf3 server init fails'
++ echo 'ERROR: iperf3 server init fails'
ERROR: iperf3 server init fails
++ exit 1
+ local result=
+ parse_iperf3_pps 'network pps cnt cnt' ''
+ local 'TEST_NAME=network pps cnt cnt'
+ local result=
+ '[' -z '' ']'
+ die 'no result output'
+ msg='no result output'
+ echo 'ERROR: no result output'
ERROR: no result output
+ exit 1
```

I've not figured out why - I suspect either a timing thing bringing up the server or maybe the output from the docker build interferes with our parsing somewhere (but tbh I don't think that is the case).
Could you see if you can reproduce and diagnose?

**Comment by GabyCT** [comment]
@grahamwhaley, I ran it again enabling the `set -x` and it is working, here it is part of the output (I did not put it the complete output)
```
===== starting test [network pps cnt cnt] =====
++ cmd=("docker")
++ check_cmds docker
++ req_cmds=("$@")
++ local cmd req_cmds
++ for cmd in '"${req_cmds[@]}"'
++ command -v docker
++ echo 'command: docker: yes'
command: docker: yes
++ clean_env
+++ docker ps -q
++ containers_running=
++ '[' '!' -z '' ']'
++ kill_processes_before_start
+++ docker ps -q
++ DOCKER_PROCS=
++ [[ -n '' ]]
+++ check_processes /usr/bin/qemu-lite-system-x86_64
+++ process=/usr/bin/qemu-lite-system-x86_64
+++ [[ -z /usr/bin/qemu-lite-system-x86_64 ]]
+++ pgrep -f /usr/bin/qemu-lite-system-x86_64
+++ '[' 1 -eq 0 ']'
++ result=
++ [[ '' -ne 0 ]]
+++ check_processes /usr/libexec/kata-containers/kata-shim
+++ process=/usr/libexec/kata-containers/kata-shim
+++ [[ -z /usr/libexec/kata-containers/kata-shim ]]
+++ pgrep -f /usr/libexec/kata-containers/kata-shim
+++ '[' 1 -eq 0 ']'
++ result=
++ [[ '' -ne 0 ]]
+++ check_processes /usr/libexec/kata-containers/kata-proxy
+++ process=/usr/libexec/kata-containers/kata-proxy
+++ [[ -z /usr/libexec/kata-containers/kata-proxy ]]
+++ pgrep -f /usr/libexec/kata-containers/kata-proxy
+++ '[' 1 -eq 0 ']'
++ result=
++ [[ '' -ne 0 ]]
++ local 'server_extra_args= -p 5201:5201'
+++ start_server local-iperf3 'export TMPDIR=/dev/shm && iperf3 -s' ' -p 5201:5201'
+++ local image=local-iperf3
+++ local 'cmd=export TMPDIR=/dev/shm && iperf3 -s'
+++ local 'extra_args= -p 5201:5201'
++++ docker run -p 5201:5201 -d --runtime kata-runtime local-iperf3 sh -c 'export TMPDIR=/dev/shm && iperf3 -s'
+++ instance_id=be38377742df0acd511457462333d011891e3c9105d221479bade7d172cd3049
++++ docker inspect --format '{{.NetworkSettings.IPAddress}}' be38377742df0acd511457462333d011891e3c9105d221479bade7d172cd3049
+++ server_address=172.17.0.2
+++ echo 172.17.0.2
++ local server_address=172.17.0.2
++ '[' -z 172.17.0.2 ']'
++ check_iperf3_server
++ local retry=6
++ local period=0.5
++ local 'test_cmd=iperf3 -c 172.17.0.2 -t 1'
++ cmds=("netstat")
++ local cmds
++ check_cmds netstat
++ req_cmds=("$@")
++ local cmd req_cmds
++ for cmd in '"${req_cmds[@]}"'
++ command -v netstat
++ echo 'command: netstat: yes'
command: netstat: yes
++ '[' 1 ']'
++ bash -c 'iperf3 -c 172.17.0.2 -t 1'
++ echo 'iperf3 server is up!'
iperf3 server is up!
++ break
+++ netstat -atun
+++ grep 5201
+++ grep LISTEN
++ lport='tcp6       0      0 :::5201                 :::*                    LISTEN     '
++ '[' -z 'tcp6       0      0 :::5201                 :::*                    LISTEN     ' ']'
++ local 'client_command=export TMPDIR=/dev/shm && iperf3 -J -u -c 172.17.0.2 -l 64 -b 0  -t 30'
+++ start_client local-iperf3 'export TMPDIR=/dev/shm && iperf3 -J -u -c 172.17.0.2 -l 64 -b 0  -t 30' ''
+++ local image=local-iperf3
+++ local 'cmd=export TMPDIR=/dev/shm && iperf3 -J -u -c 172.17.0.2 -l 64 -b 0  -t 30'
+++ local extra_args=
++++ docker run --runtime kata-runtime local-iperf3 sh -c 'export TMPDIR=/dev/shm && iperf3 -J -u -c 172.17.0.2 -l 64 -b 0  -t 30'
...
```
You need `iperf3` to be installed in the host to run these tests.

**Comment by grahamwhaley** [comment]
@GabyCT - did you run it many times? Sorry if I was not clear enough - when I said 'Not every time,' - I meant that it works for me sometimes, but not all the time. It was a sporadic failure, but I did not figure out what made it happen or not. All running on the same shell on the same machine.

**Comment by GabyCT** [comment]
@grahamwhaley yes, I ran it 4 times building the image.

**Comment by grahamwhaley** [comment]
OK, thanks. Let me see if I can have a look at it on my local system tomorrow and figure out what is upsetting it sometimes.

**Comment by GabyCT** [comment]
@grahamwhaley , I am doing the changes to save the raw results and in that PR , I am also increasing the number of retries to start the server so maybe that will help in the problem that you describe

**Comment by GabyCT** [comment]
@grahamwhaley , yes that was the issue, I am working in another PR where I will fix that (mainly because that is not related to the Dockerfile)

**Comment by GabyCT** [comment]
@grahamwhaley the issue that you found is already solved at https://github.com/kata-containers/tests/pull/571/files#diff-ba2c5337ebff9cf413c32ecb6a4419feR218

**Comment by caoruidong** [comment]
LGTM

**Comment by chavafg** [comment]
lgtm

**Pull request merged**